### PR TITLE
Implement the DataUpdateCoordinator

### PR DESCRIPTION
**Pull request by tetienne**
See #1

### CODE REVIEW

File: custom_components/tahoma/TahomaDataUpdateCoordinator.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        self.devices: Dict[str, Device] = {d.deviceurl: d for d in devices}
+        self.listener_id = listener_id
+
+    async def _async_update_data(self) -> Dict[str, Device]:
+        """Fetch data from Tahoma."""
+        events = await self.client.fetch_event_listener(self.listener_id)
+        for event in events:
+            if event.name == "DeviceStateChangedEvent":
+                for state in event.device_states:
+                    self.devices[event.deviceurl].states[state.name].value = state.value
```
Comment (tetienne) [inline custom_components/tahoma/TahomaDataUpdateCoordinator.py:47]: TODO: cast according to the returned type.

**Comment by tetienne** [comment]
I don't have fix the climate platform yet. @vlebourl I think I will need your help.

File: custom_components/tahoma/__init__.py [hunk 2]
```diff
@@ -40,17 +44,30 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
...
+        client=client,
+        devices=await client.get_devices(),
+        listener_id=await client.register_event_listener(),
+        update_interval=timedelta(seconds=5),
+    )
+
+    await coordinator.async_refresh()
+
     hass.data[DOMAIN][entry.entry_id] = {
         "controller": client,
```
Comment (iMicknl) [inline custom_components/tahoma/__init__.py:63]: Could be removed? Looks not necessary anymore :).

File: custom_components/tahoma/__init__.py [hunk 3]
```diff
@@ -40,17 +44,30 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
...
 
+    coordinator = TahomaDataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="Tahoma fetcher",
+        client=client,
+        devices=await client.get_devices(),
+        listener_id=await client.register_event_listener(),
+        update_interval=timedelta(seconds=5),
```
Comment (iMicknl) [inline custom_components/tahoma/__init__.py:57]: Is every 5 seconds not a bit to fast? Currently we only poll for updates when a device changes. This is great and will allow us to update the state if something happens outside HA.

But is every 30 seconds not enough? And maybe poll every second when we expect an update from a component, until we have received the new device state.

File: custom_components/tahoma/TahomaDataUpdateCoordinator.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from datetime import timedelta
```
Comment (iMicknl) [inline custom_components/tahoma/TahomaDataUpdateCoordinator.py:1]: Is the name of this file according to the python guidelines? Can we maybe just call it coordinator.py?

File: custom_components/tahoma/__init__.py [hunk 5]
```diff
@@ -1,15 +1,19 @@
...
 
+import async_timeout
 from tahoma_api.client import TahomaClient
 from tahoma_api.exceptions import BadCredentialsException, TooManyRequestsException
 
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, EVENT_HOMEASSISTANT_STOP
 from homeassistant.core import HomeAssistant
 
+from ...helpers.update_coordinator import DataUpdateCoordinator
```
Comment (iMicknl) [inline custom_components/tahoma/__init__.py:15]: Should be removed? It is breaking my build at least.

File: custom_components/tahoma/light.py [hunk 6]
```diff
@@ -36,10 +36,11 @@ async def async_setup_entry(hass, entry, async_add_entities):
     """Set up the TaHoma lights from a config entry."""
```
Comment (iMicknl) [inline custom_components/tahoma/light.py:36]: Light has the same issue as cover.

```
return int(brightness * 255 / 100)
TypeError: unsupported operand type(s) for /: 'str' and 'int'
```

File: custom_components/tahoma/__init__.py [hunk 7]
```diff
@@ -40,17 +44,30 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
...
         return False
 
-    devices = await client.get_devices()
     scenes = await client.get_scenarios()
 
+    coordinator = TahomaDataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="Tahoma fetcher",
```
Comment (iMicknl) [inline custom_components/tahoma/__init__.py:53]:
```suggestion
        name="TaHoma Event Fetcher",
```

File: custom_components/tahoma/climate_st.py [hunk 8]
```diff
@@ -21,6 +23,7 @@
 from homeassistant.core import callback
 from homeassistant.helpers.event import async_track_state_change
 
+from . import TahomaDataUpdateCoordinator
```
Comment (iMicknl) [inline custom_components/tahoma/climate_st.py:26]:
```suggestion
from .coordinator import TahomaDataUpdateCoordinator
```

File: custom_components/tahoma/__init__.py (hunk 3 above)
Comment (tetienne) [reply custom_components/tahoma/__init__.py:55]: Yes 5 seconds was for testing, I will set it to 30 seconds. About the second remark, we can see this in a second PR. This one is already big enough.

File: custom_components/tahoma/__init__.py (hunk 2 above)
Comment (tetienne) [reply custom_components/tahoma/__init__.py:61]: You are right.

File: custom_components/tahoma/light.py (hunk 6 above)
Comment (tetienne) [reply custom_components/tahoma/light.py:36]: Fixed.

File: custom_components/tahoma/climate_st.py (hunk 8 above)
Comment (tetienne) [reply custom_components/tahoma/climate_st.py]: Done.

File: custom_components/tahoma/binary_sensor.py [hunk 9]
```diff
@@ -61,10 +60,10 @@
 async def async_setup_entry(hass, entry, async_add_entities):
     """Set up the TaHoma sensors from a config entry."""
     data = hass.data[DOMAIN][entry.entry_id]
-    controller = data.get("controller")
+    coordinator = data.get("coordinator")
 
     entities = [
-        TahomaBinarySensor(device, controller)
+        TahomaBinarySensor(device.url, coordinator)
```
Comment (iMicknl) [inline custom_components/tahoma/binary_sensor.py:66]: device.url?

File: custom_components/tahoma/tahoma_device.py [hunk 10]
```diff
@@ -1,14 +1,12 @@
...
 
-from tahoma_api.client import TahomaClient
-from tahoma_api.models import Command, Device
+from tahoma_api.models import Command
 
 from homeassistant.const import ATTR_BATTERY_LEVEL
 from homeassistant.helpers.entity import Entity
-from homeassistant.util import Throttle
 
+from . import TahomaDataUpdateCoordinator
```
Comment (iMicknl) [inline custom_components/tahoma/tahoma_device.py:9]:
```suggestion
from .coordinator import TahomaDataUpdateCoordinator
```

Comment (tetienne) [reply custom_components/tahoma/tahoma_device.py:9]: That's isort that do it. I cannot change it.

File: custom_components/tahoma/coordinator.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+            hass, logger, name=name, update_interval=update_interval,
+        )
+
+        self.client = client
+        self.devices: Dict[str, Device] = {d.deviceurl: d for d in devices}
+        self.listener_id = listener_id
+
+    async def _async_update_data(self) -> Dict[str, Device]:
+        """Fetch data from Tahoma."""
+        events = await self.client.fetch_event_listener(self.listener_id)
```
Comment (iMicknl) [inline custom_components/tahoma/coordinator.py:46]: Should there be no try catch here? I see all integrations and samples use `            raise UpdateFailed()`. We can also have exceptions here which we should handle.

**Comment by iMicknl** [comment]
https://github.com/iMicknl/ha-tahoma/blob/1aa76f03fa17ba49e319418f809ba33a0c2cd43d/custom_components/tahoma/__init__.py#L108

Here we should make use of the right client, this one is not existing anymore.

File: custom_components/tahoma/__init__.py (hunk 3 above)
Comment (iMicknl) [reply custom_components/tahoma/__init__.py]: @tetienne @vlebourl maybe we should keep 5 seconds for now? Until we switch to an approach where we update once every x minutes and every second after an update. 

If we set it to 30 seconds, updates will still be slow.

File: custom_components/tahoma/binary_sensor.py (hunk 9 above)
Comment (tetienne) [reply custom_components/tahoma/binary_sensor.py:66]: You are right. I think I will create a PR request to rename device.deviceurl to device.url, make more sense and we don't repeat ourself.

File: custom_components/tahoma/coordinator.py (hunk 11 above)
Comment (tetienne) [reply custom_components/tahoma/coordinator.py]: Done.

**Comment by tetienne** [comment]
> https://github.com/iMicknl/ha-tahoma/blob/1aa76f03fa17ba49e319418f809ba33a0c2cd43d/custom_components/tahoma/__init__.py#L108
> 
> Here we should make use of the right client, this one is not existing anymore.

Good catch.

File: custom_components/tahoma/__init__.py (hunk 3 above)
Comment (tetienne) [reply custom_components/tahoma/__init__.py]: I'm afraid 5 seconds will overlead the HA, it's recommended to keep high. I was not able to find a component where such low poll interval.

About your suggestion to reduce it, when we just do an update using HA, we can see if we can schedule an update 5s later.

**Pull request merged**